src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MtAmazonDynamoDbStreamsBySharedTable.java

```diff
@@ -43,30 +44,108 @@ public GetShardIteratorResult getShardIterator(GetShardIteratorRequest request)
-        Predicate<MtRecord> recordFilter) {
+    protected MtGetRecordsResult getRecords(GetRecordsRequest request, Function<Record, MtRecord> recordMapper,
+                                            Predicate<MtRecord> recordFilter) {
         Optional.ofNullable(request.getLimit()).ifPresent(limit -> checkArgument(limit > 0 && limit <= MAX_LIMIT));
 
-        List<Record> mtRecords = new ArrayList<>(Optional.ofNullable(request.getLimit()).orElse(MAX_LIMIT));
-        String iterator = request.getShardIterator();
+        final int limit = Optional.ofNullable(request.getLimit()).orElse(MAX_LIMIT);
+        final long timeLimit = mtDynamoDb.getGetRecordsTimeLimit();
+        final long time = mtDynamoDb.getClock().millis();
 
-        // keep fetching records until we reach the (current) end or find at least one record for this tenant
-        do {
-            GetRecordsResult result = dynamoDbStreams.getRecords(request.clone().withShardIterator(iterator));
-            iterator = result.getNextShardIterator();
-            List<Record> records = result.getRecords();
-            if (records.isEmpty()) {
-                break;
-            }
-            records.stream().map(recordMapper).filter(recordFilter).collect(toCollection(() -> mtRecords));
-        } while (mtRecords.isEmpty() && iterator != null);
+        final MtGetRecordsResult result = new MtGetRecordsResult()
+            .withRecords(new ArrayList<>(limit))
+            .withNextShardIterator(request.getShardIterator())
+            .withLastSequenceNumber(null);
+
+        while (addRecords(result, limit, recordMapper, recordFilter)
+            && result.getRecords().size() < limit
+            && result.getNextShardIterator() != null
+            && (mtDynamoDb.getClock().millis() - time) <= timeLimit) {
+        }
 
-        return new GetRecordsResult().withRecords(mtRecords).withNextShardIterator(iterator);
+        return result;
     }
 
+    /**
+     * Fetches more records using the next iterator in the result and adds them to the result records. Returns whether
+     * to keep going or not.
+     */
+    private boolean addRecords(MtGetRecordsResult result,
+                               int limit,
+                               Function<Record, MtRecord> recordMapper,
+                               Predicate<MtRecord> recordFilter) {
+        // retrieve max number of records from underlying stream (and retry below if we got too many)
+        final GetRecordsRequest innerRequest = new GetRecordsRequest().withLimit(MAX_LIMIT)
+            .withShardIterator(result.getNextShardIterator());
+        GetRecordsResult innerResult = dynamoDbStreams.getRecords(innerRequest);
+        result.setNextShardIterator(innerResult.getNextShardIterator());
+        List<Record> records = innerResult.getRecords();
+        // if we did not any records, no point continuing
+        if (records.isEmpty()) {
```

Comment:
there could be a NPE if `MtGetRecordsResult().withRecords()` is never called (based on records implementation in aws GetRecordsResult).  This is not the case for `getRecords` but just thinking ahead if MtGetRecordsResult is reused.

Might be helpful to add a default records value in MtGetRecordsResult, before doing operations on records?

Reply:
I am not quite sure I follow. This line calls a method on the result of calling `innerResult.getRecords()` which is returned by the underlying `dynamoDbStreams` implementation. How is this related to `MtGetRecordsResult`?

Reply:
Looks like I confused the two in the comment (records and input `MtGetRecordsResult` result). I was observing that if `MtGetRecordsResult: withRecords` is called and the input parameter has a value of `null`, then `MtGetRecordsResult: withRecords` calls aws `GetRecordsResult:setRecords` implementation and in this case it will set the `records` to `null`.  It's only if `GetRecordsResult:withRecords` is called, that a non null default is set.  It doesn't affect the implementation in `addRecords`, it is just something to keep in mind , which is not to pass in a null collection into `MtGetRecordsResult:withRecords`, since we're overriding the withRecords method, and calling `GetRecordsResult:setRecords` directly.  Was suggesting we handle null input param cases in `MtGetRecordsResult:withRecords`.

Reply:
GetRecordsResult:setRecords:   
 public void setRecords(Collection<Record> records) {
        if (records == null) {
            **this.records = null;**
        } else {
            this.records = new ArrayList(records);
        }
    }

Reply:
`GetRecordsResult` only defaults to an empty collection for `withRecords(Record... records)`, not for `withRecords(java.util.Collection<Record> records)`, correct?

Reply:
then in `MtAmazonDynamoDbStreamsBySharedTable:addRecords(MtGetRecordsResult result,..)` 
`final int innerLimit = limit - result.getRecords().size();` , if records are null, result.getRecords() would be null

Reply:
`result#records` is initialized to an empty list on line 67. How could `result.getRecords()` be null? Or are you just talking purely from the perspective of this method?

I think in general, I am assuming an invariant that records are never null, because either our code or the sdk initializes it to a (possibly empty) collection. You make a good point though, that maybe we should be programming more defensively?

Reply:
yes `result#records` is initialized to an empty list on line 67 , so it's handled for this merge.  Mainly commenting on `MtGetRecordsResult: withRecords` usage, and I noticed this case when reviewing  `MtAmazonDynamoDbStreamsBySharedTable:addRecords(MtGetRecordsResult result,..)`, so we're good for this, but since `MtGetRecordsResult: withRecords` is public, a user of mt-dynamo who calls `MtGetRecordsResult: withRecords` could run into this if input is `null`.

Yes, from what I saw in aws's implementation, they don't default initialize unless `GetRecordsResult withRecords(Record... records)` which is the concern since we are overriding the other `withRecords` method that sets null to null.

Reply:
The request and result objects in the AWS SDK are just DTOs. There are no cases where clients build a result object and pass it into the dynamo sdk; it's always the sdk building a result object and returning it to clients. You could argue that they shouldn't expose setters and maybe even make them immutable objects, but presumably they had their reasons for this approach. It does make things like unit testing (mocking) easier.

mt-dynamo follows the same pattern: it exposes a mutable `MtGetRecordsResult` object, but only exposes methods that return instances of it (none that take it as an input argument).